### PR TITLE
 Add find machines mode 

### PR DESCRIPTION
**Pull request by CodingBenny**
I added an option to show the machines a recipe can be crafted in.

### CODE REVIEW

File: crafting_combinator/script/rc.lua [hunk 1]
```diff
@@ -22,6 +22,16 @@ function _M.init_global()
...
+	local crafting_machines = {}
+	for name,prototype in pairs(game.entity_prototypes) do
+		if prototype.crafting_categories and game.forces[1].recipes[name] then
+			for categorie in pairs(prototype.crafting_categories) do
+				crafting_machines[categorie] = crafting_machines[categorie] or {}
+				table.insert(crafting_machines[categorie],name)
+			end
+		end
+	end
+	global.rc.machines = crafting_machines
```
Comment (theRustyKnife) [inline crafting_combinator/script/rc.lua]: Is there a particular reason this needs to be in global? I'd prefer if it was just a local that gets rebuilt when necessary.

File: crafting_combinator/control.lua [hunk 2]
```diff
@@ -51,6 +51,7 @@ script.on_load(on_load)
 script.on_configuration_changed(function(changes)
 	update_bottleneck(changes)
 	enable_recipes()
+	rc_control.init_global()
```
Comment (theRustyKnife) [inline crafting_combinator/control.lua]: This really shouldn't be called here, it's intended only for the initial setup. If there's any setup to do on version changes, put that into a one-time migration instead. Though with my other suggestion, this shouldn't be necessary at all.

File: crafting_combinator/script/rc.lua [hunk 3]
```diff
@@ -140,13 +155,37 @@ function _M:find_ingredients_and_products(forced)
...
+
+	if self.recipe ~= recipe or forced then
+		self.recipe = recipe
+		local params = {}
+		if recipe and recipe.category then
+			for i, mac in pairs(global.rc.machines[recipe.category] or {}) do
+				local mac_res = self.entity.force.recipes[mac]
+				if mac_res and mac_res.enabled then
+					table.insert(params, {
+						signal = recipe_selector.get_signal(mac),
```
Comment (theRustyKnife) [inline crafting_combinator/script/rc.lua]: This should use different logic than the recipes. Again, it should probably be the item that the machine is built with.

File: crafting_combinator/script/rc.lua [hunk 4]
```diff
@@ -140,13 +155,37 @@ function _M:find_ingredients_and_products(forced)
...
 
+function _M:find_machines(forced)
+	local recipe, input_count = recipe_selector.get_recipe(self.entity, nil, defines.circuit_connector_id.combinator_input)
+
+	if self.recipe ~= recipe or forced then
+		self.recipe = recipe
+		local params = {}
+		if recipe and recipe.category then
+			for i, mac in pairs(global.rc.machines[recipe.category] or {}) do
+				local mac_res = self.entity.force.recipes[mac]
```
Comment (theRustyKnife) [inline crafting_combinator/script/rc.lua]: This wouldn't work for machines that have a different name than the recipe that produces them. This should probably check if there is some recipe that produces the item that places the machine instead.

File: crafting_combinator/script/rc.lua [hunk 5]
```diff
@@ -140,13 +155,37 @@ function _M:find_ingredients_and_products(forced)
 end
 
 
+function _M:find_machines(forced)
+	local recipe, input_count = recipe_selector.get_recipe(self.entity, nil, defines.circuit_connector_id.combinator_input)
+
+	if self.recipe ~= recipe or forced then
```
Comment (theRustyKnife) [inline crafting_combinator/script/rc.lua]: This should probably use the same condition as [here](https://github.com/theRustyKnife/CraftingCombinator/blob/10d307dcb87577feaedc96db3e7a10289f6da6ec/crafting_combinator/script/rc.lua#L107) to update when the input count changes.

**Review by theRustyKnife — CHANGES_REQUESTED** [review]
This is nice, and something that people have been asking me for, so it's great that you did this. There are a few things that need revisions tho.

Also, my inner grammar nazi urges me to ask you to fix the spelling of "category" in init_global :P

File: crafting_combinator/script/rc.lua (hunk 5 above)
Comment (CodingBenny) [reply crafting_combinator/script/rc.lua]: I had that first, but I was messing around with it when i tried to update it after a new machine was researched.

File: crafting_combinator/script/rc.lua (hunk 3 above)
Comment (CodingBenny) [reply crafting_combinator/script/rc.lua]: It probably does not matter if we output the signal of the item used to place to machine or the recipe you get the item from. I am outputing the item now.

**Comment by CodingBenny** [comment]
I added an option to improve signal splitting: If the strength of the signal is the same as the index, users can easily split the signals by using decider combinators that only passes a "each = 1" or "each = 2" or "each = A" signal on.

**Review by theRustyKnife — APPROVED** [review]

**Comment by theRustyKnife** [comment]
I quite like the idea of numbering the output, it'll definitely come in handy. Tho please open a new PR for unrelated changes next time.

Anyway, this looks fine, so I'll merge it.

**Pull request merged**